website/server/models/user/methods.js

```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
   return userGroups;
 };
 
+// Get an array of error message keys that would be thrown if the given interaction was attempted
+schema.methods.getObjectionsToInteraction = function getObjectionsToInteraction (interaction, receiver) {
+  let sender = this;
+
+  /* eslint-disable no-unused-vars */ // The checks below all get access to sndr and rcvr, but not all use both
+  let checks = {
+    always: [
+      // Revoked chat privileges block all interactions to prevent the evading of harassment protections
+      // See issue #7971 for some discussion
+      (sndr, rcvr) => sndr.flags.chatRevoked && 'chatPrivilegesRevoked',
+
+      // Direct user blocks prevent all interactions
+      (sndr, rcvr) => rcvr.inbox.blocks.includes(sndr._id) && 'notAuthorizedToSendMessageToThisUser',
+      (sndr, rcvr) => sndr.inbox.blocks.includes(rcvr._id) && 'notAuthorizedToSendMessageToThisUser',
+    ],
+
+    'send-private-message': [
+      // Private messaging has an opt-out, which does not affect other interactions
+      (sndr, rcvr) => rcvr.inbox.optOut && 'notAuthorizedToSendMessageToThisUser',
+
+      // We allow a player to message themselves so they can test how PMs work or send their own notes to themselves
+    ],
+
+    'transfer-gems': [
+      // Unlike private messages, gems can't be sent to oneself
+      (sndr, rcvr) => rcvr._id === sndr._id && 'cannotSendGemsToYourself',
+    ],
+  };
+  /* eslint-enable no-unused-vars */
+
+  let knownInteractions = Object.keys(checks).filter((key) => key !== 'always');
```

Comment:
The `checks` and `knownInteractions` variables can be pull out of this function since they are constant and do not need to be recreated for every time this method is called. 

Maybe:

```js
const INTERACTION_CHECKS = Object.freeze({ /* functions here */ });
const KNOWN_INTERACTIONS = Object.freeze(Object.keys(INTERACTION_CHECKS).filter(key => key !== 'always'));

This means you can also export `KNOWN_INTERACTIONS` for a params check in the route instead of throwing an error here.
```